cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaMetadataStateManager.java

```diff
@@ -0,0 +1,397 @@
+/*
+ * Copyright Strimzi authors.
+ * License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
+ */
+package io.strimzi.operator.cluster.operator.resource;
+
+import io.fabric8.kubernetes.api.model.Secret;
+import io.strimzi.api.kafka.model.Kafka;
+import io.strimzi.operator.cluster.model.KafkaCluster;
+import io.strimzi.operator.common.Annotations;
+import io.strimzi.operator.common.Reconciliation;
+import io.strimzi.operator.common.ReconciliationLogger;
+
+import java.util.Locale;
+
+/**
+ * Class used to reconcile the metadata state which represents where metadata are stored (ZooKeeper or KRaft)
+ * It is also used to compute metadata state changes during the migration process from ZooKeeper to KRaft (or rollback)
+ */
+public class KafkaMetadataStateManager {
+
+    private static final ReconciliationLogger LOGGER = ReconciliationLogger.create(KafkaMetadataStateManager.class.getName());
+
+    /**
+     * The annotation value which indicates that the ZooKeeper to KRaft migration is enabled
+     */
+    public static final String MIGRATION_VALUE_STRIMZI_IO_KRAFT = "migration";
```

Comment:
I was actually thinking the other way around. At some point in the future, we won't have migration anymore and only KRaft based cluster. The `KafkaMetadataStateManager` class will be removed and with it all the related metadata stuff, because it will be all in KRaft.

Reply:
Right, let's put them all together in KafkaMetadataStateManager.